src/com/trilead/ssh2/Connection.java

```diff
@@ -672,65 +672,67 @@ public synchronized ConnectionInfo connect(ServerHostKeyVerifier verifier, int c
 		{
 			TimeoutToken token = null;
 
-			if (kexTimeout > 0)
+			try
 			{
-				final Runnable timeoutHandler = new Runnable()
+				if (kexTimeout > 0)
 				{
-					public void run()
+					final Runnable timeoutHandler = new Runnable()
 					{
-						synchronized (state)
+						public void run()
 						{
-							if (state.isCancelled)
-								return;
-							state.timeoutSocketClosed = true;
-							tm.close(new SocketTimeoutException("The connect timeout expired"), false);
+							synchronized (state)
+							{
+								if (state.isCancelled)
+									return;
+								state.timeoutSocketClosed = true;
+								tm.close(new SocketTimeoutException("The connect timeout expired"), false);
```

Comment:
Would likely be a good idea to call directly close() (from this call) here instead of tm.close() to avoid a NPE. close() does many extra checks and eventually calls tm.close().

Reply:
probably make sense, in one of the catch make it